### PR TITLE
[REFACTOR] replace Map with std::map

### PR DESCRIPTION
**Pull request by timosachsenberg**
# Description
Steps necessary:
- [x] Changed interface to std::map for access
- [x] Remove .has from Map interface and use modern constructs like:
```
if (auto m_it = mymap.find(key); m_it != mymap.end())
{
 // do something with m_it
}

to replace code like:
if (mymap.has(key))
{
  // do something with mymap.at(key)
}
```
- [x] replace Map with std::map C++
- [x] update python bindings / conversion providers

# Checklist:
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes. (Tick if no updates were necessary.)

# How can I get additional information on failed tests during CI:
If your PR is failing you can check out 
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. If you click in the column that lists the failed tests you will get detailed error messages.
- Or click on the action: e.g., for clang-format linting

# Note:
- Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

# Advanced commands (admins / reviewer only):
- `/rebase` will try to rebase the PR on the current develop branch.
- `/reformat` (experimental) applies the clang-format style changes as additional commit
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds

### CODE REVIEW

**Comment by timosachsenberg** [comment]
rebuild jenkins

**Comment by timosachsenberg** [comment]
hi, can anyone hold hands for the last mile? Python bindings and conversion providers are the last parts missing

**Comment by jpfeuffer** [comment]
Let's just try find-replace Map with libcpp_map (including its include). Then we will see if anything made an assumption on the specific Map functions.

**Comment by jpfeuffer** [comment]
We don't need any conversion providers anymore. If at all, we should remove the one for Map.

File: src/openms/source/ANALYSIS/QUANTITATION/ItraqConstants.cpp [hunk 1]
```diff
@@ -209,7 +209,7 @@ namespace OpenMS
       info.description = "";
       info.name = channel_names[itraq_type].getValue(i, 0);
       info.id = (Int)i;
-      if (!reporter_mass_exact.has(info.name))
+      if (!(reporter_mass_exact.find(info.name) != reporter_mass_exact.end()))
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/QUANTITATION/ItraqConstants.cpp:212]: Please replace check for non-existence with the simpler:
if (reporter_mass_exact.find(info.name) == reporter_mass_exact.end())

**Comment by hroest** [comment]
> Let's just try find-replace Map with libcpp_map (including its include). Then we will see if anything made an assumption on the specific Map functions.

yes that would work, the only thing we will have to keep in mind is the Map special conversion providers which we need to adopt

https://github.com/OpenMS/OpenMS/blob/58ac2727f251f14febb1e2ddc063472f4113ee43/src/pyOpenMS/converters/special_autowrap_conversionproviders.py#L620

**Comment by hroest** [comment]
with b19fa09aa01cb4c0b2f2e7db1df688e7dd8b9f4b now pyOpenMS compiles and all tests run. However I have not tested TOPP tools and unit tests

**Comment by hroest** [comment]
PS: as  6aeecd7096d5d8e14b9b05d27681ae6993d39416 shows, we need to go over this by hand with a fine comb and cant just automatically replace things, if we create a map temporarily than we cannot compare `end()` of two different temporary maps. This just shows that this can lead to some subtle bugs...

File: src/openms/source/ANALYSIS/DECHARGING/ILPDCWrapper.cpp [hunk 2]
```diff
@@ -97,13 +98,13 @@ namespace OpenMS
             g2f.erase(group2);
           }
         }
-        else if (f2g.has(f1) && !f2g.has(f2)) // only f1 is part of a group
+        else if (f2g.find(f1) != f2g.end() && !(f2g.find(f2) != f2g.end())) // only f1 is part of a group
```
Comment (hroest) [inline src/openms/source/ANALYSIS/DECHARGING/ILPDCWrapper.cpp:101]:
```suggestion
        else if ((f2g.find(f1) != f2g.end()) && !(f2g.find(f2) != f2g.end())) // only f1 is part of a group
```

File: src/openms/source/ANALYSIS/DECHARGING/ILPDCWrapper.cpp [hunk 3]
```diff
@@ -72,15 +73,15 @@ namespace OpenMS
...
+      std::map<Size, Size> f2g; // feature id to connected group
+      std::map<Size, std::set<Size> > g2pairs; // group id to all pairs involved
+      std::map<Size, std::set<Size> > g2f; // group id to all features involved
 
       for (Size i = 0; i < pairs.size(); ++i)
       {
         Size f1 = pairs[i].getElementIndex(0);
         Size f2 = pairs[i].getElementIndex(1);
-        if (f2g.has(f1) && f2g.has(f2)) // edge connects two distinct groups
+        if (f2g.find(f1) != f2g.end() && f2g.find(f2) != f2g.end()) // edge connects two distinct groups
```
Comment (hroest) [inline src/openms/source/ANALYSIS/DECHARGING/ILPDCWrapper.cpp:84]:
```suggestion
        if ((f2g.find(f1) != f2g.end()) && (f2g.find(f2) != f2g.end())) // edge connects two distinct groups
```

**Comment by hroest** [comment]
@timosachsenberg how far away are we from C++20 and std::map::contains ? https://en.cppreference.com/w/cpp/container/map/contains

**Comment by timosachsenberg** [comment]
C20 support is still incomplete in most compilers but such simpler methods are well supported. VS caught up a bit though. Major point speaking against an upgrade is support by conda afaik. @jpfeuffer right?

**Comment by jpfeuffer** [comment]
Conda is at GCC 9 I think. If that supports it, I think it would be fine.
I think we should target around Visual Studio 2019. And maybe MacOS Big Sur or the one before. Not sure about their compatibilities right now.

File: src/openms/source/ANALYSIS/QUANTITATION/ItraqConstants.cpp (hunk 1 above)
Comment (timosachsenberg) [reply src/openms/source/ANALYSIS/QUANTITATION/ItraqConstants.cpp:212]: @matteopilz

**Comment by timosachsenberg** [comment]
Some python errors remain: https://abibuilder.informatik.uni-tuebingen.de/jenkins/job/openms/job/PR/job/pyOMSDckr/2404/console
otherwise passes

**Comment by matteopilz** [comment]
Judging from the tests, the last error seems to come from https://github.com/OpenMS/OpenMS/blob/develop/src/pyOpenMS/tests/unittests/test000.py.
However, I cannot find out whats wrong.

File: src/pyOpenMS/addons/ADD_TO_ALL.pyx [hunk 4]
```diff
@@ -1,7 +1,6 @@
 
 from libc.stddef cimport ptrdiff_t
 from libc.stdint cimport *
```
Comment (hroest) [inline src/pyOpenMS/addons/ADD_TO_ALL.pyx:3]:
```suggestion
from libc.stdint cimport *
from libcpp.map cimport map as libcpp_map
```

**Comment by hroest** [comment]
> Some python errors remain: https://abibuilder.informatik.uni-tuebingen.de/jenkins/job/openms/job/PR/job/pyOMSDckr/2404/console

seems like a failure to compile, therefore all tests fail

**Comment by matteopilz** [comment]
> > Some python errors remain: https://abibuilder.informatik.uni-tuebingen.de/jenkins/job/openms/job/PR/job/pyOMSDckr/2404/console
> 
> seems like a failure to compile, therefore all tests fail

Only two tests are failing now: https://abibuilder.informatik.uni-tuebingen.de/jenkins/job/openms/job/PR/job/pyOMSDckr/2408/console
test_pyopenms_unittests & pyopenms_unittest_test000
However, I am unable to find the issue here.

**Comment by timosachsenberg** [comment]
having the output uploaded would be of great help

**Comment by jpfeuffer** [comment]
Yep. There is probably just the upload command in the cmake script missing. You can also upload from the command line.

**Comment by timosachsenberg** [comment]
@matteopilz can you check if you can add the upload command?

**Comment by matteopilz** [comment]
> @matteopilz can you check if you can add the upload command?

Where do I insert the command?

**Comment by jpfeuffer** [comment]
We are not using any scripts for the pyopenms build so you need to either do everything from the command line (not sure how to get all the setting from the CTestConfig.cmake inside there but probably possible).
Or you need to write a Cmake script like we do for GitHub actions.

**Comment by jpfeuffer** [comment]
Click on details, go back to the job overview, configure, search for the bash script.

**Comment by matteopilz** [comment]
[pyopenms_unittest_test000_stacktrace.txt](https://github.com/OpenMS/OpenMS/files/8204948/pyopenms_unittest_test000_stacktrace.txt)
[pyopenms_unittest_test000_std_output.txt](https://github.com/OpenMS/OpenMS/files/8204949/pyopenms_unittest_test000_std_output.txt)

This is the decompressed output from: https://abibuilder.informatik.uni-tuebingen.de/jenkins/job/openms/job/PR/job/pyOMSDckr/2408/testReport/projectroot/pyOpenMS/pyopenms_unittest_test000_py/

**Comment by jpfeuffer** [comment]
So something wrong with AAseq getMonoWeight.
Either in the function or with the construction/retrieval of seq

File: src/openms/thirdparty/evergreen/src/Utility/graph_to_dot.hpp [hunk 5]
```diff
@@ -2,6 +2,8 @@
 #define _GRAPH_TO_DOT_HPP
 
 #include <fstream>
+#include <map>
```
Comment (jpfeuffer) [inline src/openms/thirdparty/evergreen/src/Utility/graph_to_dot.hpp:5]: What happened here?

Comment (matteopilz) [reply src/openms/thirdparty/evergreen/src/Utility/graph_to_dot.hpp:5]: What do you mean?

Comment (timosachsenberg) [reply src/openms/thirdparty/evergreen/src/Utility/graph_to_dot.hpp:5]: I think he means that the thidparty library should not have any changes regarding the map

Comment (matteopilz) [reply src/openms/thirdparty/evergreen/src/Utility/graph_to_dot.hpp]: I've changed it back now.

**Comment by matteopilz** [comment]
unittest still fails

**Comment by hroest** [comment]
I quickly tested this, its not a Python issue. In C++ this also fails:

```
    AASequence seq = AASequence::fromString("PEPTIDESEKUEM(Oxidation)CER");
    TEST_EQUAL(seq.toUniModString(), "PEPTIDESEKUEM(UniMod:35)CER")
    TEST_REAL_SIMILAR(seq.getMonoWeight(), 1952.7200317517998)
```

```
371: checking AASequence fromString(const String& s, bool permissive = true) ...                                                                                                          371:  -  line 294:  TEST_REAL_SIMILAR(seq.getMonoWeight(),1952.7200317517998): got 1958.7140766518, expected 1952.7200317518 (absolute: 5.99404490000006 [1e-05], relative: 1.00306958744855 [1.00001], message: "ratio of numbers is large"                                                                                                                                        371: : failed                                                                                                                                                                             371:                            
```

**Comment by hroest** [comment]
its probably this bug: #5904 and has nothing to do with this PR. We can wait for this to resolve first and then merge?

File: src/openms/include/OpenMS/DATASTRUCTURES/sources.cmake [hunk 6]
```diff
@@ -31,8 +31,7 @@ KDTree.h
 ListUtils.h
 ListUtilsIO.h
 LPWrapper.h
-Map.h
-MassExplainer.h
+        MassExplainer.h
```
Comment (hroest) [inline src/openms/include/OpenMS/DATASTRUCTURES/sources.cmake:34]:
```suggestion
MassExplainer.h
```

File: src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h [hunk 7]
```diff
@@ -47,6 +47,7 @@
 #include <OpenMS/CONCEPT/LogStream.h>
 #include <OpenMS/CONCEPT/ProgressLogger.h>
 #include <vector>
+#include <vector>
```
Comment (hroest) [inline src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h:50]:
```suggestion
```

**Review by hroest — APPROVED** [review]
It looks all good, all tests pass!

**Pull request merged**